modules/broker-core/src/main/java/org/wso2/broker/core/store/dao/impl/BaseDao.java

```diff
@@ -0,0 +1,131 @@
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package org.wso2.broker.core.store.dao.impl;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.broker.common.util.function.ThrowingConsumer;
+import org.wso2.broker.common.util.function.ThrowingFunction;
+import org.wso2.broker.core.BrokerException;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Objects;
+import javax.sql.DataSource;
+
+/**
+ * Implements methods to do boilerplate JDBC tasks.
+ */
+abstract class BaseDao {
+    private static final Logger LOGGER = LoggerFactory.getLogger(BaseDao.class);
+
+    private final DataSource dataSource;
+
+    BaseDao(DataSource dataSource) {
+        this.dataSource = dataSource;
+    }
+
+    Connection getConnection() throws SQLException {
+        return dataSource.getConnection();
+    }
+
+    private void close(Connection connection) {
+        if (connection != null) {
+            try {
+                connection.close();
+            } catch (SQLException e) {
+                LOGGER.error("Error closing connection.", e);
+            }
+        }
+    }
+
+    void close(Connection connection, Statement statement) {
+        close(statement);
+        close(connection);
+    }
+
+    void close(Connection connection, Statement statement, ResultSet resultSet) {
+        close(resultSet);
+        close(connection, statement);
+    }
+
+    void close(Statement statement) {
+        if (statement != null) {
+            try {
+                statement.close();
+            } catch (SQLException e) {
+                LOGGER.error("Error closing statement. " + statement, e);
+            }
+        }
+    }
+
+    void close(ResultSet resultSet) {
+        if (resultSet != null) {
+            try {
+                resultSet.close();
+            } catch (SQLException e) {
+                LOGGER.error("Error closing result set.", e);
+            }
+        }
+    }
+
+    void rollback(Connection connection, String message) {
+        try {
+            if (Objects.nonNull(connection)) {
+                connection.rollback();
+            }
+        } catch (SQLException e) {
+            LOGGER.error("Error occurred while rolling back. Failed operation " + message, e);
+        }
+    }
+
+    <E extends Exception> void transaction(ThrowingConsumer<Connection, E> command,
```

Comment:
Can't we use the following signature instead?
`void transaction(ThrowingConsumer<Connection, Exception> command,`

Reply:
Updated.